"use strict";(self.webpackChunkpop_gen_jl=self.webpackChunkpop_gen_jl||[]).push([[4567],{3905:function(e,a,t){t.d(a,{Zo:function(){return d},kt:function(){return m}});var n=t(7294);function o(e,a,t){return a in e?Object.defineProperty(e,a,{value:t,enumerable:!0,configurable:!0,writable:!0}):e[a]=t,e}function p(e,a){var t=Object.keys(e);if(Object.getOwnPropertySymbols){var n=Object.getOwnPropertySymbols(e);a&&(n=n.filter((function(a){return Object.getOwnPropertyDescriptor(e,a).enumerable}))),t.push.apply(t,n)}return t}function r(e){for(var a=1;a<arguments.length;a++){var t=null!=arguments[a]?arguments[a]:{};a%2?p(Object(t),!0).forEach((function(a){o(e,a,t[a])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(t)):p(Object(t)).forEach((function(a){Object.defineProperty(e,a,Object.getOwnPropertyDescriptor(t,a))}))}return e}function l(e,a){if(null==e)return{};var t,n,o=function(e,a){if(null==e)return{};var t,n,o={},p=Object.keys(e);for(n=0;n<p.length;n++)t=p[n],a.indexOf(t)>=0||(o[t]=e[t]);return o}(e,a);if(Object.getOwnPropertySymbols){var p=Object.getOwnPropertySymbols(e);for(n=0;n<p.length;n++)t=p[n],a.indexOf(t)>=0||Object.prototype.propertyIsEnumerable.call(e,t)&&(o[t]=e[t])}return o}var i=n.createContext({}),s=function(e){var a=n.useContext(i),t=a;return e&&(t="function"==typeof e?e(a):r(r({},a),e)),t},d=function(e){var a=s(e.components);return n.createElement(i.Provider,{value:a},e.children)},u={inlineCode:"code",wrapper:function(e){var a=e.children;return n.createElement(n.Fragment,{},a)}},c=n.forwardRef((function(e,a){var t=e.components,o=e.mdxType,p=e.originalType,i=e.parentName,d=l(e,["components","mdxType","originalType","parentName"]),c=s(t),m=o,f=c["".concat(i,".").concat(m)]||c[m]||u[m]||p;return t?n.createElement(f,r(r({ref:a},d),{},{components:t})):n.createElement(f,r({ref:a},d))}));function m(e,a){var t=arguments,o=a&&a.mdxType;if("string"==typeof e||o){var p=t.length,r=new Array(p);r[0]=c;var l={};for(var i in a)hasOwnProperty.call(a,i)&&(l[i]=a[i]);l.originalType=e,l.mdxType="string"==typeof e?e:o,r[1]=l;for(var s=2;s<p;s++)r[s]=t[s];return n.createElement.apply(null,r)}return n.createElement.apply(null,t)}c.displayName="MDXCreateElement"},2120:function(e,a,t){t.r(a),t.d(a,{frontMatter:function(){return l},contentTitle:function(){return i},metadata:function(){return s},assets:function(){return d},toc:function(){return u},default:function(){return m}});var n=t(7462),o=t(3366),p=(t(7294),t(3905)),r=["components"],l={id:"types",title:"PopData.jl",sidebar_label:"PopData.jl"},i=void 0,s={unversionedId:"api/PopGenCore/types",id:"api/PopGenCore/types",title:"PopData.jl",description:"PopGenCore.jl/src/PopData.jl",source:"@site/docs/api/PopGenCore/Types.md",sourceDirName:"api/PopGenCore",slug:"/api/PopGenCore/types",permalink:"/PopGen.jl/docs/api/PopGenCore/types",editUrl:"https://github.com/BioJulia/PopGen.jl/edit/documentation/docs/api/PopGenCore/Types.md",tags:[],version:"current",lastUpdatedAt:1636029729,formattedLastUpdatedAt:"11/4/2021",frontMatter:{id:"types",title:"PopData.jl",sidebar_label:"PopData.jl"},sidebar:"docs",previous:{title:"Structure.jl",permalink:"/PopGen.jl/docs/api/PopGenCore/structure"},next:{title:"VariantCall.jl",permalink:"/PopGen.jl/docs/api/PopGenCore/variantcall"}},d={},u=[{value:"PopGenCore.jl/src/PopData.jl",id:"popgencorejlsrcpopdatajl",level:2},{value:"\ud83d\udfea\ud83d\udd35 PopObj",id:"-popobj",level:3},{value:"\ud83d\udfea\ud83d\udd35 PopDataInfo",id:"-popdatainfo",level:3},{value:"\ud83d\udfea\ud83d\udd35 PopData",id:"-popdata",level:3},{value:"\ud83d\udfea\ud83d\udd35 PopDataInfo!",id:"-popdatainfo-1",level:3},{value:"\ud83d\udfea\ud83d\udd35 Genotype",id:"-genotype",level:3},{value:"\ud83d\udfea\ud83d\udd35 SNP",id:"-snp",level:3},{value:"\ud83d\udce6 _SNP",id:"-_snp",level:3},{value:"\ud83d\udfea\ud83d\udd35 MSat",id:"-msat",level:3},{value:"\ud83d\udce6 _MSat",id:"-_msat",level:3},{value:"\ud83d\udfea\ud83d\udd35 GenoArray",id:"-genoarray",level:3},{value:"\ud83d\udce6 _ploidy2text",id:"-_ploidy2text",level:3},{value:"\ud83d\udfea\ud83d\udd35 Base.show",id:"-baseshow",level:3},{value:"\ud83d\udfea\ud83d\udd35 Base.getindex",id:"-basegetindex",level:3},{value:"\ud83d\udfea\ud83d\udd35 Base.getproperty",id:"-basegetproperty",level:3}],c={toc:u};function m(e){var a=e.components,t=(0,o.Z)(e,r);return(0,p.kt)("wrapper",(0,n.Z)({},c,t,{components:a,mdxType:"MDXLayout"}),(0,p.kt)("h2",{id:"popgencorejlsrcpopdatajl"},"PopGenCore.jl/src/PopData.jl"),(0,p.kt)("p",null,"\ud83d\udce6  => not exported |\n\ud83d\udfea => exported by PopGenCore.jl |\n\ud83d\udd35 => exported by PopGen.jl"),(0,p.kt)("h3",{id:"-popobj"},"\ud83d\udfea\ud83d\udd35 PopObj"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-Julia"},"PopObj::AbstractType\n")),(0,p.kt)("p",null,"Generic AbstractType for use in PopGen.jl"),(0,p.kt)("h3",{id:"-popdatainfo"},"\ud83d\udfea\ud83d\udd35 PopDataInfo"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre"},"mutable struct PopDataInfo\n    samples::Int64\n    sampeinfo::DataFrame\n    loci::Int64\n    locusinfo::DataFrame\n    populations::Int64\n    ploidy::Union{Int8, Vector{Int8}}\n    biallelic::Bool\nend\n")),(0,p.kt)("p",null,"The data struct used internally as ",(0,p.kt)("inlineCode",{parentName:"p"},"PopData.metadata")," fields to store basic information\nabout the ",(0,p.kt)("inlineCode",{parentName:"p"},"PopData")," for easy access."),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"PopDataInfo(genodf::DataFrame)\n")),(0,p.kt)("p",null,"constructor format using just the genodata dataframe"),(0,p.kt)("h3",{id:"-popdata"},"\ud83d\udfea\ud83d\udd35 PopData"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre"},"PopData\n    metadata::PopDataInfo\n    genodata::DataFrame\n")),(0,p.kt)("p",null,"The data struct used for the PopGen population genetics ecosystem. You are\n",(0,p.kt)("strong",{parentName:"p"},"strongly")," discouraged from manually creating tables to pass into a PopData,\nand instead should use the provided file importers and utilities."),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre"},"- `metadata` PopDataInfo of  data information\n    - `samples` - the number of samples in the data\n    - `sampleinfo` - DataFrame of sample names,populations, ploidy, etc.\n    - `loci` - the number of loci in the data\n    - `locusinfo` - DataFrame of locus names, chromosome, physical position, etc.\n    - `populations` - the number of populations in the data\n    - `ploidy` - the ploidy (or ploidies) present in the data\n    - `biallelic` - if all the markers are biallelic\n- `genodata` DataFrame of sample genotype records\n    - `name` - the individual/sample names [`PooledArray`]\n    - `population` - population names [`PooledArray`]\n    - `locus` - locus names [`PooledArray`]\n    - `genotype` - genotype values [`NTuple{N,Signed}`]\n")),(0,p.kt)("p",null,(0,p.kt)("inlineCode",{parentName:"p"},"PopData(data::DataFrame)"),"\nContructor using just a ",(0,p.kt)("inlineCode",{parentName:"p"},"genodata")," dataframe."),(0,p.kt)("h3",{id:"-popdatainfo-1"},"\ud83d\udfea\ud83d\udd35 PopDataInfo!"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"PopDataInfo!(data::PopData)\n")),(0,p.kt)("p",null,"This method is used to update PopDataInfo from PopData, all in one swoop"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"PopDataInfo!(popdatainfo::PopDataInfo, genodata::DataFrame)\n")),(0,p.kt)("p",null,"Method to update preexisting PopDataInfo with new genodata, useful for getindex and creating new PopData from that"),(0,p.kt)("h3",{id:"-genotype"},"\ud83d\udfea\ud83d\udd35 Genotype"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"Genotype::DataType\n")),(0,p.kt)("p",null,"For convenience purposes, an alias for ",(0,p.kt)("inlineCode",{parentName:"p"},"NTuple{N, <:Signed} where N"),", which is\nthe type describing individual genotypes in PopData. Specifically, there exist\n",(0,p.kt)("inlineCode",{parentName:"p"},"SNP")," as an alias for ",(0,p.kt)("inlineCode",{parentName:"p"},"NTuple{N, Int8}")," and ",(0,p.kt)("inlineCode",{parentName:"p"},"MSat")," for ",(0,p.kt)("inlineCode",{parentName:"p"},"NTuple{N, Int16}")),(0,p.kt)("h3",{id:"-snp"},"\ud83d\udfea\ud83d\udd35 SNP"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"SNP::DataType\n")),(0,p.kt)("p",null,"An alias for ",(0,p.kt)("inlineCode",{parentName:"p"},"NTuple{N, Int8}")),(0,p.kt)("h3",{id:"-_snp"},"\ud83d\udce6 _SNP"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"SNP(geno)\n")),(0,p.kt)("p",null,"Contstructor for ",(0,p.kt)("inlineCode",{parentName:"p"},"SNP")),(0,p.kt)("h3",{id:"-msat"},"\ud83d\udfea\ud83d\udd35 MSat"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"MSat::DataType\n")),(0,p.kt)("p",null,"An alias for ",(0,p.kt)("inlineCode",{parentName:"p"},"NTuple{N, Int16}")),(0,p.kt)("h3",{id:"-_msat"},"\ud83d\udce6 _MSat"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"_MSat(geno)\n")),(0,p.kt)("p",null,"Constructor for ",(0,p.kt)("inlineCode",{parentName:"p"},"MSat")),(0,p.kt)("h3",{id:"-genoarray"},"\ud83d\udfea\ud83d\udd35 GenoArray"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"GenoArray::DataType\n")),(0,p.kt)("p",null,"An alias for an ",(0,p.kt)("inlineCode",{parentName:"p"},"AbstractVector")," of elements ",(0,p.kt)("inlineCode",{parentName:"p"},"Missing"),"\nand ",(0,p.kt)("inlineCode",{parentName:"p"},"Genotype"),", which itself is of type ",(0,p.kt)("inlineCode",{parentName:"p"},"NTuple{N, <:Integer} where N"),".\nThe definition as an ",(0,p.kt)("inlineCode",{parentName:"p"},"AbstractVector")," adds flexibility for ",(0,p.kt)("inlineCode",{parentName:"p"},"SubArray"),"\ncases."),(0,p.kt)("h3",{id:"-_ploidy2text"},"\ud83d\udce6 _ploidy2text"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre"},"_ploidy2text(ploidy::Int8)\n_ploidy2text(ploidy::Vector{Int8})\n")),(0,p.kt)("h3",{id:"-baseshow"},"\ud83d\udfea\ud83d\udd35 Base.show"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"Base.show(io::IO, data::PopData)\nBase.show(io::IO, data::PopDataInfo)\n")),(0,p.kt)("h3",{id:"-basegetindex"},"\ud83d\udfea\ud83d\udd35 Base.getindex"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"Base.getindex(data::PopData, idx::Symbol)\nBase.getindex(data::PopData, args)\nBase.getindex(data::PopData, expression, cols)\n")),(0,p.kt)("h3",{id:"-basegetproperty"},"\ud83d\udfea\ud83d\udd35 Base.getproperty"),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"getproperty(data::PopData, field::Symbol)\n")),(0,p.kt)("p",null,"A convenience method to access certain elements in a ",(0,p.kt)("inlineCode",{parentName:"p"},"PopData")," with fewer keystrokes.\nEssentially a standard ",(0,p.kt)("inlineCode",{parentName:"p"},"getproperty")," call, except ",(0,p.kt)("inlineCode",{parentName:"p"},"sampleinfo")," accesses ",(0,p.kt)("inlineCode",{parentName:"p"},"metadata.sampleinfo"),",\n",(0,p.kt)("inlineCode",{parentName:"p"},"locusinfo")," accesses ",(0,p.kt)("inlineCode",{parentName:"p"},"metadata.locusinfo"),", and ",(0,p.kt)("inlineCode",{parentName:"p"},"info")," is an alias for ",(0,p.kt)("inlineCode",{parentName:"p"},"metadata"),"."),(0,p.kt)("p",null,(0,p.kt)("strong",{parentName:"p"},"Example")),(0,p.kt)("pre",null,(0,p.kt)("code",{parentName:"pre",className:"language-julia"},"cats = @nancycats ;\ncats.metadata == cats.info\ncats.metadata.sampleinfo == cats.sampleinfo\ncats.metadata.locusinfo == cats.locusinfo\n")))}m.isMDXComponent=!0}}]);